model/models/nomicbert/model.go

```diff
@@ -0,0 +1,179 @@
+	Up   *nn.Linear `gguf:"ffn_up"`
+	Down *nn.Linear `gguf:"ffn_down"`
+}
+
+// Forward implements model.Model interface
+func (m *Model) Forward(ctx ml.Context, batch input.Batch) (ml.Tensor, error) {
+	seqLen := batch.Inputs.Dim(0)
+
+	hiddenStates := m.TokenEmbedding.Forward(ctx, batch.Inputs)
+
+	typeEmbed := m.TypeEmbedding.Weight.View(ctx, 0, m.hiddenSize)
+	hiddenStates = hiddenStates.Add(ctx, typeEmbed)
+
+	// no position embedding - we use ROPE instead
+
+	hiddenStates = m.TokenEmbeddingNorm.Forward(ctx, hiddenStates, m.eps)
+
+	positions := ctx.Arange(0, float32(seqLen), 1, ml.DTypeI32)
+
+	for _, layer := range m.Layers {
+		hiddenStates = layer.Forward(ctx, hiddenStates, positions, &m.Options)
+	}
+
+	hiddenStates = m.poolingType.Forward(ctx, hiddenStates)
+
+	if m.normalize {
+		hiddenStates = hiddenStates.L2Norm(ctx, 1e-12)
+	}
+
+	return hiddenStates, nil
+}
+
+func (e *EncoderLayer) Forward(ctx ml.Context, hiddenStates ml.Tensor, positions ml.Tensor, opts *Options) ml.Tensor {
+	residual := hiddenStates
+	hiddenStates = e.Attention.Forward(ctx, hiddenStates, positions, opts)
+	hiddenStates = hiddenStates.Add(ctx, residual)
+	hiddenStates = e.AttentionNorm.Forward(ctx, hiddenStates, opts.eps)
+
+	residual = hiddenStates
+	hiddenStates = e.MLP.Forward(ctx, hiddenStates)
+	hiddenStates = hiddenStates.Add(ctx, residual)
+	hiddenStates = e.MLPNorm.Forward(ctx, hiddenStates, opts.eps)
+
+	return hiddenStates
+}
+
+func (a *Attention) Forward(ctx ml.Context, hiddenStates ml.Tensor, positions ml.Tensor, opts *Options) ml.Tensor {
+	batchSize := hiddenStates.Dim(1)
+
+	qkv := a.QKV.Forward(ctx, hiddenStates)
+
+	query := qkv.View(ctx, 0, opts.hiddenSize, qkv.Stride(0), batchSize).Contiguous(ctx)
+	key := qkv.View(ctx, opts.hiddenSize*qkv.Stride(0), opts.hiddenSize, qkv.Stride(0), batchSize).Contiguous(ctx)
+	value := qkv.View(ctx, 2*opts.hiddenSize*qkv.Stride(0), opts.hiddenSize, qkv.Stride(0), batchSize).Contiguous(ctx)
+
+	query = query.Reshape(ctx, opts.headDim, opts.numHeads, batchSize)
+	key = key.Reshape(ctx, opts.headDim, opts.numHeads, batchSize)
+	value = value.Reshape(ctx, opts.headDim, opts.numHeads, batchSize)
+
+	query = fast.RoPE(ctx, query, positions, opts.headDim, opts.ropeFreqBase, 1.0)
+	key = fast.RoPE(ctx, key, positions, opts.headDim, opts.ropeFreqBase, 1.0)
+
+	attention := nn.Attention(ctx, query, key, value, 1.0/math.Sqrt(float64(opts.headDim)), nil)
+
+	attention = attention.Reshape(ctx, opts.hiddenSize, batchSize)
+
+	return a.Output.Forward(ctx, attention)
+}
+
+func (m *MLP) Forward(ctx ml.Context, hiddenStates ml.Tensor) ml.Tensor {
+	gate := m.Gate.Forward(ctx, hiddenStates).SILU(ctx)
+	up := m.Up.Forward(ctx, hiddenStates)
+	hidden := gate.Mul(ctx, up)
+
+	return m.Down.Forward(ctx, hidden)
+}
+
+func New(c fs.Config) (model.Model, error) {
+	hiddenSize := int(c.Uint("embedding_length"))
+	numHeads := int(c.Uint("attention.head_count"))
+	headDim := hiddenSize / numHeads
+
+	processor := model.NewWordPiece(
```

Comment:
this model doesn't want its inputs lowercased. there's a todo in `WordPiece.Encode(...)` to handle this case. you should make that change as well